### PR TITLE
spi generic layer + adc1 example

### PR DESCRIPTION
**Pull request by amiclaus**
This a proposal for the spi generic layer implementation.

And example of usage is provided using the fmcjesdadc1 project with the
Xilinx platform specific driver.

An additional layer will be added afterwards between spi generic layer
and xilinx spi layer for the CPLD demux spi driver.

Signed-off-by: Antoniu Miclaus <antoniu.miclaus@analog.com>

### CODE REVIEW

**Comment by amiclaus** [comment]
Appropiate indentation and comments will be added when the pull request is made.

**Comment by mchindri** [comment]
[spi_proposal.txt](https://github.com/analogdevicesinc/no-OS/files/4461942/spi_proposal.txt)

**Comment by mchindri** [comment]
```
//spi.h
struct spi_interface {
	int32_t (*spi_init)(struct spi_desc **, const struct spi_init_param *);
	int32_t (*spi_write_and_read)(struct spi_desc *, uint8_t *, uint16_t);
	int32_t (*spi_remove)(struct spi_desc *);
}

struct spi_init_param {
	struct spi_interface *interface;
	//...
}

struct spi_desc {
	struct spi_interface *interface;
	//...
}

//xilinx/spi_extra.h
int32_t xil_configure_interface(struct spi_interface *interface);

//xilinx/spi.c
int32_t xil_configure_interface(struct spi_interface *interface) {
	interface->spi_init = xil_spi_init;
	interface->spi_write_and_read = xil_spi_write_and_read;
	interface->spi_remove = xili_spi_remove;
}

//spi.c
int32_t spi_init(struct spi_desc **desc,
		 const struct spi_init_param *param){

	int32_t ret;

	ret = param->interface->spi_init(desc, param);
	if (IS_ERR_VALUE(ret))
		return (ret);
	//Here de descriptor is allocated by the driver
	desc->interface = param->interface;

	return SUCCESS;
}

//fmcjesdadc1.c
struct spi_desc *ad9250;
struct spi_init_param ad9250_spi_param = {
	//...
}
xil_configure_interface(&ad9250_spi_param.spi_interface);
spi_init(&ad9250, &ad9250_spi_param);

```

**Comment by dbogdan** [comment]
I like the idea of having an interface, but having to call an additional function (configure_interface()) is not ideal... Maybe each driver can directly expose the interface to be used by spi_init().

File: include/spi.h [hunk 1]
```diff
@@ -72,6 +72,11 @@ typedef enum spi_mode {
 	SPI_MODE_3 = (SPI_CPOL | SPI_CPHA)
 } spi_mode;
 
+struct spi_dev {
```
Comment (scuciurean) [inline include/spi.h:75]: Maybe find a better fitting naming? Like spi_ops or spi_actions.

**Comment by mchindri** [comment]
@dbogdan You mean an exposed interface like this? It seems better to me too.
`
//xilinx/spi_extra.h
constr struct spi_interface xil_spi_interface = {
	.spi_init = xil_spi_init,
	.spi_write_and_read = xil_spi_init,
	.spi_remove = xili_spi_remove
}
`

**Comment by dbogdan** [comment]
@mchindri , yes, something like that.

File: drivers/spi/spi.c [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+	free(desc->dev);
+
+	return SUCCESS;
+}
+
+/* Write and read data to/from SPI. */
+int32_t spi_write_and_read(struct spi_desc *desc,
+			   uint8_t *data,
+			   uint16_t bytes_number){
+	return (desc->dev->spi_dev_write_and_read(desc, data, bytes_number));
```
Comment (buha) [inline drivers/spi/spi.c:40]: no-OS is a library, test for NULL function pointer

Comment (buha) [reply drivers/spi/spi.c:40]: or maybe even better, return some failure code in `spi_init`, the driver doesn't make sense to lack any of these pointers (maybe some drivers could lack the `remove`)

this way you dont need to test for NULL here

**Comment by adrimbarean** [comment]
How many projects do have this problem? I could argue that the SPI driver and the device drivers are fine as it is, but it's actually the applications responsibility here to provide correct SPI implementation that will replace the one in the driver. I don't think we need to change too much xilinx spi driver. Just enough to make it accept custom implementation if necessary.

**Comment by dbogdan** [comment]
Besides the SPI expander used on ADC1, all the SPI engine projects are affected by this limitation. What we did until now (considering the SPI engine as a Xilinx SPI type) was only a workaround. SPI engine should be treated as an independent SPI controller since it's available on Intel platforms too.

**Comment by amiclaus** [comment]
Closing this draft.
PR added with #577
